lib/appsignal/tesla.ex

```diff
@@ -0,0 +1,188 @@
+        :ok ->
+          _ = Appsignal.IntegrationLogger.debug("Appsignal.Tesla attached to #{inspect(event)}")
+
+          :ok
+
+        {:error, _} = error ->
+          Logger.warn("Appsignal.Tesla not attached to #{inspect(event)}: #{inspect(error)}")
+
+          error
+      end
+    end
+  end
+
+  def tesla_request_start(
+        _event,
+        _measurements,
+        %{env: env},
+        _config
+      ) do
+    do_tesla_request_start(@tracer.current_span(), env)
+  end
+
+  defp do_tesla_request_start(nil, _env), do: nil
+
+  defp do_tesla_request_start(parent, env) do
+    %{method: method, url: url} = env
+
+    middlewares = env_middlewares(env)
+    base_url = middleware_base_url(middlewares)
+    use_full_path = path_params?(middlewares)
+
+    sanitised_url = sanitise_url(url, base_url, use_full_path)
+
+    upcased_method = String.upcase(Atom.to_string(method))
+
+    name = "#{upcased_method} #{sanitised_url}"
+
+    "http_request"
+    |> @tracer.create_span(parent)
+    |> @span.set_name(name)
+    |> @span.set_attribute("appsignal:category", "request.tesla")
+  end
+
+  # Sanitises an URL by keeping only its scheme, host and port.
+  #
+  # If `use_full_path` is true, the path fragment is also kept.
+  # The `use_full_path` parameter is set to true for enhanced grouping
+  # when `Tesla.Middleware.PathParams` is present.
+  #
+  # If the URL does not contain a host, and a base URL is provided by
+  # `Tesla.Middleware.BaseUrl`, then the base URL's scheme, host,
+  # port and path are used instead.
+  # When using the base URL, its path fragment will be used, regardless
+  # of the value of `use_full_path`.
+  # If `use_full_path` is true, then the path fragments of the base URL
+  # and the URL are concatenated together.
+  def sanitise_url(url, base_url, use_full_path) do
+    %URI{
+      scheme: scheme,
+      host: host,
+      port: port,
+      path: path
+    } = URI.parse(url || "")
+
+    # Users can specify a host in the request's URL, which overrides the
+    # base URL in the BaseURL middleware. Only use the base URL if the URL
+    # in the telemetry event does not already have a host.
+    # If the BaseURL middleware is used before the Telemetry middleware,
+    # the telemetry event's URL will already have the base URL interpolated
+    # into it, and so the base URL argument will never be used.
+    if host == nil do
+      %URI{
+        scheme: base_scheme,
+        host: base_host,
+        port: base_port,
+        path: base_path
+      } = URI.parse(base_url || "")
+
+      full_path =
+        if use_full_path do
+          "#{base_path}#{path}"
+        else
+          base_path
+        end
+
+      URI.to_string(%URI{scheme: base_scheme, host: base_host, port: base_port, path: full_path})
+    else
+      full_path = if use_full_path, do: path
+
+      URI.to_string(%URI{scheme: scheme, host: host, port: port, path: full_path})
+    end
+  end
+
+  def env_middlewares(env) do
```

Comment:
I don't think this function is used anywhere else. Could you make it private?

Reply:
It's public so that the unit tests can access it, same as `sanitise_url`. I thought that otherwise, testing the edge cases of this function by sending events through Telemetry got a bit unwieldy.